harness/determined/pytorch/deepspeed/_deepspeed_trial.py

```diff
@@ -0,0 +1,880 @@
+            if self.context.mpu.should_report_metrics():
+                # Verify validation metric names are the same across batches.
+                if keys is None:
+                    keys = vld_metrics.keys()
+                else:
+                    if keys != vld_metrics.keys():
+                        raise det.errors.InvalidExperimentException(
+                            "Validation metric names must match across all batches of data.",
+                        )
+                if not isinstance(vld_metrics, dict):
+                    raise det.errors.InvalidExperimentException(
+                        "validation_metrics() must return a "
+                        "dictionary of string names to Tensor "
+                        "metrics",
+                    )
+                # TODO: For performance perform -> cpu() only at the end of validation.
+                batch_metrics.append(pytorch._convert_metrics_to_numpy(vld_metrics))
+            if self.env.test_mode:
+                break
+
+        all_keys = self.context.distributed._zmq_gather(keys and list(keys))
+        if self.is_chief:
+            all_keys = [k for k in all_keys if k]
+            keys = all_keys[0]
+        keys = self.context.distributed._zmq_broadcast(keys)
+
+        for callback in self.callbacks.values():
+            callback.on_validation_epoch_end(batch_metrics)
+
+        metrics = pytorch._reduce_metrics(
+            self.context.distributed,
+            batch_metrics=batch_metrics,
+            keys=keys,
+            metrics_reducers=pytorch._prepare_metrics_reducers(pytorch.Reducer.AVG, keys=keys),
+        )
+        metrics.update(
+            pytorch._convert_metrics_to_numpy(self.context.reduce_metrics(for_training=False))
+        )
+
+        if self.context.distributed.size > 1 and any(
+            map(
+                lambda c: util.is_overridden(c.on_validation_end, pytorch.PyTorchCallback)
+                or util.is_overridden(c.on_validation_step_end, pytorch.PyTorchCallback),
+                self.callbacks.values(),
+            )
+        ):
+            logging.debug(
+                "Broadcasting metrics to all worker processes to execute a "
+                "validation step end callback"
+            )
+            metrics = self.context.distributed._zmq_broadcast(metrics)
+
+        for callback in self.callbacks.values():
+            if util.is_overridden(callback.on_validation_step_end, pytorch.PyTorchCallback):
+                logging.warning(
+                    "on_validation_step_end is now deprecated, please use on_validation_end instead"
+                )
+                callback.on_validation_step_end(metrics)
+
+        for callback in self.callbacks.values():
+            callback.on_validation_end(metrics)
+
+        if not self.is_chief:
+            return {}
+
+        num_inputs *= self.context.mpu.get_data_parallel_world_size()
```

Comment:
Can we make this more correct?  I recently [added](https://github.com/determined-ai/determined/pull/3593/files#diff-31d6b7cf29d7c9ddaa96513a19ddbda78c2005cc7c6fdcb275248702871fc9beR562) actual allgathers of validation data sizes to pytorch trial.  At least with pytorchtrial, it's especially important now that `make_timing_log` prints the number of actual inputs them model saw; it shouldn't look like we added extra inputs to the dataset somehow.

Reply:
Since we are passing an iterator over the dataset to `train_batch` and `evaluate_batch`, I didn't see a good way to get the actual number of samples, especially for pipeline parallel which calls next on the iterator multiple times per forward call.  Even in DeepSpeed, the builtin throughput timer makes the same assumption that all batches are of size `train_micro_batch_size_per_gpu` https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/engine.py#L269